docs: add openshift install

ericpaulsen opened the pull request:
initial draft that outlines the steps (and quirks) of deploying Coder on OpenShift.

I think we should add a brief section at the bottom for using the `BuildConfig` to build images that are compliant with SCC UID restrictions. we could also show how those images can be referenced in a terraform template.

please add/clarify anything I missed.

fixes #4646

ericpaulsen committed:
docs: add openshift install

bpmct commented on docs/install/openshift.md:
```diff
@@ -0,0 +1,148 @@
+## Requirements
+
+Before proceeding, please ensure that you have an OpenShift cluster running K8s
+1.19+ (OpenShift 4.7+) and have Helm 3.5+ installed. In addition, you'll need to
+install the OpenShift CLI (`oc`) to authenticate to your cluster and create OpenShift
+resources.
```
Is there a reason why customers need this version installed? It could take a customer months to get the proper cluster, so we may want to clarify whether this is a hard requirement or simply what we have tested Coder with.

ericpaulsen replied:
this is just keeping in line with our K8s install docs, which say you need a cluster with version 1.19+ (the OpenShift equivalent is 4.7).

bpmct commented on docs/install/openshift.md:
````diff
@@ -0,0 +1,148 @@
+Depending upon your configured Security Context Constraints (SCC), you'll need to set
+the following `securityContext` values in the Coder Helm chart:
+
+```yaml
+coder:
+  securityContext:
+    runAsNonRoot: true
+    runAsUser: 1000
+    runAsGroup: 1000
+    readOnlyRootFilesystem: true
+    seccompProfile:
+      type: RuntimeDefault
+    allowPrivilegeEscalation: false
+```
````
If the user builds their own `coder` image with a different UID, will they need to `allowPrivilegeEscalation` and set `runAsUser/runAsGroup`?

bpmct commented on docs/install/openshift.md:
```diff
@@ -0,0 +1,148 @@
+
+### 3. Set the `CODER_CACHE_DIRECTORY` environment variable
+
+By default, Coder creates the cache directory in `/home/coder/.cache`. Given the
+OpenShift-provided UID, the Coder container does not have permission to write to
+this directory.
```
Doesn't `runAsUser: 1000` give the container user permissions to write in this directory?

sharkymark replied:
this is the Coder container, not the workspace container

bpmct replied:
Yeah, the Coder container creates a `coder` user which runs as 1000 I believe.

bpmct reviewed (commented):
It's looking great so far! I had some questions, but mostly curious. Don't consider any of them blocking.

I agree we should provide a sample buildConfig so that folks can create workspace in OpenShift.

ericpaulsen committed:
make: fmt

johnstcn commented on docs/install/openshift.md:
```diff
@@ -0,0 +1,148 @@
+    runAsNonRoot: true
+    runAsUser: 1000
+    runAsGroup: 1000
+    readOnlyRootFilesystem: true
```
```suggestion
    readOnlyRootFilesystem: false
```

johnstcn commented on docs/install/openshift.md:
```diff
@@ -0,0 +1,148 @@
+coder:
+  securityContext:
+    runAsNonRoot: true
+    runAsUser: 1000
```
```suggestion
    runAsUser: <project-specific UID>
```

johnstcn commented on docs/install/openshift.md:
```diff
@@ -0,0 +1,148 @@
+  securityContext:
+    runAsNonRoot: true
+    runAsUser: 1000
+    runAsGroup: 1000
```
```suggestion
    runAsGroup: <project-specific GID>
```

johnstcn committed:
Apply suggestions from code review

Update securityContext: set project-specific uid/gid, set readOnlyRootFilesystem=false

johnstcn committed:
add steps for creating compatible image and template

johnstcn commented on docs/install/openshift.md:
```diff
@@ -0,0 +1,260 @@
+    readOnlyRootFilesystem: false
+    seccompProfile:
+      type: RuntimeDefault
+    allowPrivilegeEscalation: false
```
```suggestion
    allowPrivilegeEscalation: false
    capabilities:
      drop:
        - ALL
```

johnstcn committed:
Update docs/install/openshift.md

sharkymark commented on docs/install/openshift.md:
````diff
@@ -0,0 +1,263 @@
+Run the following command to login to your OpenShift cluster:
+
+```console
+oc login <cluster-url>
+```
````
this just returns a message that you need an API token.

suggest we replace with:

oc login --token=sha256~<token> --server=<cluster-url>

ericpaulsen replied:
@sharkymark yes but the `oc login <cluster-url` flow gets the user to retrieve the token

sharkymark commented on docs/install/openshift.md:
````diff
@@ -0,0 +1,263 @@
+oc login <cluster-url>
+```
+
+This will configure your local `~/.kube/config` file with the cluster credentials
+needed when installing Coder via `helm`.
````
Suggest deleting the helm part - oc does lots of things like get pods, describe nodes, etc.

sharkymark commented on docs/install/openshift.md:
```diff
@@ -0,0 +1,263 @@
+
+Set the `CODER_PG_CONNECTION_URL` value to enable Coder to establish a connection
+to a PostgreSQL instance. [See our Helm documentation](./kubernetes.md) on configuring
+the PostgreSQL connection URL as a secret. Additionally, if accessing Coder over a hostname, set
+the `CODER_ACCESS_URL` value.
```
Isn't it more that say the CODER_ACCESS_URL should be set, like with any install? This comes across as optional, while technically true in some case.

ericpaulsen replied:
@sharkymark if not set in the values file, `CODER_ACCESS_URL` will automatically be set by Coder to the `cluster.local` DNS name, so we don't need to say it's a requirement, per se.

bpmct commented on docs/install/openshift.md:
````diff
@@ -0,0 +1,263 @@
+oc new-project coder
+```
+
+### 2. Configure SecurityContext values
+
+Depending upon your configured Security Context Constraints (SCC), you'll need to set
+the following `securityContext` values in the Coder Helm chart:
````
I noticed we didn't do this for our internal Openshift deployment. I'm assuming some deployments need it? 

Should we make it clear it's not a required step? And where does one find their `<project-specific UID>?`

ericpaulsen replied:
@bpmct Cian added this in the BuildConfig section.

```
oc get project coder -o json | jq -r '.metadata.annotations'
```

do you think we should add this here as well?

bpmct replied:
Yep!

bpmct replied:
In my install, we didn't have to do any of the seccompProfile, allowPriviledgeEscalation, and capability stuff. It was just

```yaml
coder:
  replicaCount: 1
  env:
    - name: CODER_PG_CONNECTION_URL
      valueFrom:
        secretKeyRef:
          key: url
          name: coder-db-url
    - name: CODER_CACHE_DIRECTORY
      value: /tmp/coder-cache/
  securityContext:
    runAsGroup: 1001080000
    runAsUser: 1001080000
  service:
    enable: true
    externalTrafficPolicy: Cluster
    sessionAffinity: ClientIP
    type: ClusterIP
```

Are there other installs that require all of the other stuff?

Also, I'm noticing that setting different sections of the helm values are different steps. What do you think about combining it into a single step and using comments or an explainer around why each section (securityContext, service type) is important. It might be a bit confusing with multiple steps on all the values that need to be set.

ericpaulsen replied:
@bpmct good feedback, i'll remove those values and also combine the Helm steps (Coder service, access URL, postgreSQL, cache stuff).

bpmct commented:
OpenShift may set low limits/requests which makes code-server and other apps fail. We should look into this!

johnstcn committed:
make fmt

ericpaulsen committed:
docs: make changes on feedback

ericpaulsen committed:
Merge branch 'main' into openshift-docs

bpmct commented on docs/install/openshift.md:
````diff
@@ -0,0 +1,259 @@
+      value: /tmp/coder-cache
+```
+
+### 4. Set access URL, PostgreSQL connection values
````
Openshift's "marketplace" offers Postgres as an option which works nicely with Coder. Mark and I left some scratch notes on this [here](https://www.notion.so/coderhq/OpenShift-Notes-a027bc331ee7495193930e037023c14e?pvs=4). I think we should mention this here. Thoughts?

ericpaulsen replied:
@bpmct i think we should - couldn't hurt.

ericpaulsen committed:
combine helm steps
